nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+	Header http.Header
+	// State of the operation, should be failed or canceled.
+	State nexusapi.OperationState
+	// Failure object to send with the completion.
+	Failure *nexusapi.Failure
+}
+
+// NewJSONSuccessfulOperationCompletion constructs an [OperationCompletionSuccessful] from a JSONable value.
+// Marhsals the provided value to JSON using [nexusapi.DefaultMarshaler].
+func NewJSONSuccessfulOperationCompletion(v any) (*OperationCompletionSuccessful, error) {
+	b, err := nexusapi.DefaultMarshaler(v)
+	if err != nil {
+		return nil, err
+	}
+
+	header := make(http.Header, 1)
+	header.Set(nexusapi.HeaderContentType, nexusapi.ContentTypeJSON)
+
+	return &OperationCompletionSuccessful{
+		Header: header,
+		Body:   io.NopCloser(bytes.NewReader(b)),
+	}, nil
+}
+
+func (c *OperationCompletionSuccessful) applyToHTTPRequest(request *http.Request, client *Client) error {
+	if c.Header != nil {
+		request.Header = c.Header.Clone()
+	}
+	request.Header.Set(nexusapi.HeaderOperationState, string(nexusapi.OperationStateSucceeded))
+	if closer, ok := c.Body.(io.ReadCloser); ok {
+		request.Body = closer
+	} else {
+		request.Body = io.NopCloser(c.Body)
+	}
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (c *OperationCompletionSuccessful) Close() error {
+	if closer, ok := c.Body.(io.Closer); ok {
+		return closer.Close()
+	}
+	return nil
+}
+
+func (c *OperationCompletionUnsuccessful) applyToHTTPRequest(request *http.Request, client *Client) error {
+	if c.Header != nil {
+		request.Header = c.Header.Clone()
+	}
+	request.Header.Set(nexusapi.HeaderOperationState, string(c.State))
+	request.Header.Set(nexusapi.HeaderContentType, nexusapi.ContentTypeJSON)
+
+	b, err := client.Options.Marshaler(c.Failure)
+	if err != nil {
+		return err
+	}
+
+	request.Body = io.NopCloser(bytes.NewReader(b))
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (c *OperationCompletionUnsuccessful) Close() error {
+	return nil
+}
+
+// DeliverCompletion delivers the result of a completed asynchronous operation to the provided URL.
+func (c *Client) DeliverCompletion(ctx context.Context, url string, completion OperationCompletion) error {
+	// while the http client is expected to close the body, we close in case request creation fails
+	defer completion.Close()
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, nil)
+	if err != nil {
+		return err
+	}
+	if err := completion.applyToHTTPRequest(httpReq, c); err != nil {
+		return err
+	}
+
+	httpReq.Header.Set(headerUserAgent, UserAgent)
+	response, err := c.Options.HTTPClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != 200 {
+		return c.newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response)
+	}
+
+	return nil
+}
+
+func (c *Client) joinURL(parts ...string) (*url.URL, error) {
+	if c.serviceBaseURL == nil {
+		return nil, ErrEmptyServiceBaseURL
+	}
```

Comment:
This should be checked at client instantiation time. And therefore this one-line utility method is not needed.